sympy/core/basic.py

```diff
@@ -1581,15 +1581,39 @@ def _eval_rewrite(self, pattern, rule, **hints):
                     return rewritten
         return self.func(*args)
 
+    def _accept_eval_derivative(self, s):
+        # This method needs to be overridden by array-like objects
+        return s._visit_eval_derivative_scalar(self)
+
+    def _visit_eval_derivative_scalar(self, base):
+        # Base is a scalar
+        # Types are (base: scalar, self: scalar)
+        return base._eval_derivative(self)
+
+    def _visit_eval_derivative_array(self, base):
+        # Types are (base: array/matrix, self: scalar)
+        # Base is some kind of array/matrix,
+        # it should have `.applyfunc(lambda x: x.diff(self)` implemented:
+        return base._eval_derivative(self)
+
     def _eval_derivative_n_times(self, s, n):
         # This is the default evaluator for derivatives (as called by `diff`
         # and `Derivative`), it will attempt a loop to derive the expression
         # `n` times by calling the corresponding `_eval_derivative` method,
         # while leaving the derivative unevaluated if `n` is symbolic.  This
         # method should be overridden if the object has a closed form for its
         # symbolic n-th derivative.
-        from sympy import Derivative
-        return Derivative._helper_apply_n_times(self, s, n, lambda x, s: x._eval_derivative(s))
+        from sympy import Integer
+        if isinstance(n, (int, Integer)):
+            obj = self
+            for i in range(n):
```

Comment:
If we have range logic here we should reject negative integers, as they will give a wrong answer (range(-n) is empty).

Reply:
Negative numbers are blocked in `Derivative`.